### PR TITLE
Add UNet resize-conv layers

### PR DESCRIPTION
**Pull request by tom-andersson**
This PR addresses the issue of [checkerboard artifacts](https://distill.pub/2016/deconv-checkerboard/) in the existing `UNet` class. The problem is addressed by using resize-convolution layers in place of transposed convolutions. A boolean argument allows for reverting back to transposed convolutions, which should help with assessing the impact of this change.

Have tested for the 2D case with TensorFlow, but will be worth testing 1D and 3D and a PyTorch backend as well.

Please review - happy to make any changes.

### CODE REVIEW

**Comment by wesselb** [comment]
This looks great, @tom-andersson! Would you be able to enable this these convs in [`test_architectures`](https://github.com/wesselb/neuralprocesses/blob/main/tests/test_architectures.py#L49)? I think it is probably best to add a new config

```
            (
                "construct_convgnp",
                {
                    "num_basis_functions": 4,
                    "points_per_unit": 16,
                    "unet_channels": (8, 16),
                    "unet_kernels": (5,) * 2,
                    "unet_activations": (B.relu,) * 2,
                    "unet_use_resizes_convs": True,  # This is added.
                    "epsilon": 1e-2,
                },
            ),
```

File: neuralprocesses/torch/nn.py [hunk 1]
```diff
@@ -101,6 +114,10 @@ def ModuleList(modules):
     Conv2d = partial(ConvNd, dim=2)
     Conv3d = partial(ConvNd, dim=3)
 
+    UpSampling1d = partial(UpSamplingNd, dim=1)
```
Comment (wesselb) [inline neuralprocesses/torch/nn.py:117]: Does this work? `UpSamplingNd` seems to not take the keyword argument `dim`.

File: neuralprocesses/torch/nn.py [hunk 2]
```diff
@@ -58,6 +58,19 @@ def ConvNd(
     )
 
 
+def UpSamplingNd(
+    size: int = 2,
+    interp_method: str = 'bilinear',
+    dtype=None,
+):
+    return getattr(torch.nn, f"Upsample")(
+        # scalar multiplier applied over each dim automatically
```
Comment (wesselb) [inline neuralprocesses/torch/nn.py:67]:
```suggestion
        # `scale_factor` is applied to each dimension automatically: it doesn't need to
        # be repeated.
```

File: neuralprocesses/torch/nn.py [hunk 3]
```diff
@@ -58,6 +58,19 @@ def ConvNd(
     )
 
 
+def UpSamplingNd(
+    size: int = 2,
+    interp_method: str = 'bilinear',
```
Comment (wesselb) [inline neuralprocesses/torch/nn.py:63]:
```suggestion
    interp_method: str = "bilinear",
```
(Black.)

File: neuralprocesses/tensorflow/nn.py [hunk 4]
```diff
@@ -85,6 +85,37 @@ def ConvNd(
         )
 
 
+def UpSamplingNd(
+    dim: int,
+    size: int = 2,
+    interp_method: str = 'bilinear',
```
Comment (wesselb) [inline neuralprocesses/tensorflow/nn.py:91]:
```suggestion
    interp_method: str = "bilinear",
```

(Black.)

**Review by wesselb — CHANGES_REQUESTED** [review]

File: neuralprocesses/torch/nn.py (hunk 1 above)
Comment (tom-andersson) [reply neuralprocesses/torch/nn.py:117]: Good catch, you're correct. PyTorch infers the dim from the shape and resizes automatically.

Comment (wesselb) [reply neuralprocesses/torch/nn.py:117]: Right! Can then simply remove the `dim=X`s I think. :)

Comment (tom-andersson) [reply neuralprocesses/torch/nn.py]: Done: https://github.com/wesselb/neuralprocesses/pull/1/commits/5963f3b60ffccf7b950a5501c74b7bda140b621c

**Comment by tom-andersson** [comment]
Thank you for the feedback @wesselb ! Regarding the test config, note that I've set the boolean argument to use the resize-convs by default, so all code/tests that don't set it will by default switch to using the resize-conv layers. I've therefore added two test configs with the `unet_uses_resize_convs` arg set to both `False` and `True` in https://github.com/wesselb/neuralprocesses/pull/1/commits/7fdab9958d5b4c978298ef1b652fee5d9f4a1fd0. Any feedback on this?

File: neuralprocesses/torch/nn.py [hunk 5]
```diff
@@ -58,6 +58,20 @@ def ConvNd(
...
+def UpSamplingNd(
+    size: int = 2,
+    interp_method: str = "bilinear",
+    dtype=None,
+):
+    return getattr(torch.nn, f"Upsample")(
+        # `scale_factor` is applied to each dimension automatically: it doesn't need to
+        # be repeated.
+        scale_factor=size,
+        method=interp_method,
```
Comment (wesselb) [inline neuralprocesses/torch/nn.py:70]:
```suggestion
        mode=interp_method,
```

**Review by wesselb — CHANGES_REQUESTED** [review]
To fix CI

File: neuralprocesses/torch/nn.py (hunk 5 above)
Comment (tom-andersson) [reply neuralprocesses/torch/nn.py:70]: Of course! Apologies, that was silly.

File: neuralprocesses/tensorflow/nn.py [hunk 6]
```diff
@@ -85,6 +85,63 @@ def ConvNd(
...
+    else:
+        data_format = "channels_last"
+
+    # Due to inconsistent input signatures between TensorFlow UpSampling
+    # classes, we must construct conditional layer args dependent
+    # on input dim
+    layer_kwargs = dict()
+    layer_kwargs["size"] = size if dim == 1 else (size,) * dim
+    layer_kwargs["dtype"] = dtype
+    if dim == 2 or dim == 3:
```
Comment (wesselb) [inline neuralprocesses/tensorflow/nn.py:106]:
```suggestion
    if dim != 1:
```

If `dim == 4`, then the logic would break. (Though, admittedly, `UpSampling4D` doesn't exist in TF, but future versions could possibly support it.)

File: neuralprocesses/architectures/convgnp.py [hunk 7]
```diff
@@ -22,6 +22,8 @@ def construct_convgnp(
         unet_channels=(64,) * 6,
         unet_activations=None,
         unet_kernels=5,
+        unet_use_resize_convs=True,
```
Comment (wesselb) [inline neuralprocesses/architectures/convgnp.py:25]: Perhaps we want to call `unet_use_resize_convs` instead `unet_resize_convs` so that it is a prefix of `unet_resize_conv_interp_method`? That looks nice aesthetically. :)

**Review by wesselb — CHANGES_REQUESTED** [review]
Only only question and one comments! Otherwise, this LGTM. :)

File: neuralprocesses/architectures/convgnp.py (hunk 7 above)
Comment (tom-andersson) [reply neuralprocesses/architectures/convgnp.py:25]: Great suggestion! I'll implement in next push.

Comment (tom-andersson) [reply neuralprocesses/architectures/convgnp.py:25]: I will also update the `UNet` arg from `use_resize_convs` to `resize_convs` to match up.
